tracing-appender/src/rolling.rs

```diff
@@ -0,0 +1,178 @@
+use crate::inner::InnerAppender;
+use chrono::{DateTime, Datelike, TimeZone, Timelike, Utc};
+use std::fmt::Debug;
+use std::fs::{File, OpenOptions};
+use std::io::{BufWriter, Write};
+use std::path::Path;
+use std::sync::{Mutex, MutexGuard};
+use std::{fs, io};
+use tracing_subscriber::fmt::MakeWriter;
+
+pub struct RollingFileAppender {
+    inner: Mutex<InnerAppender<BufWriterFactory>>,
+}
+
+impl RollingFileAppender {
+    fn writer(&self) -> RollingFileWriter {
+        RollingFileWriter::new(&self.inner)
+    }
+}
+
+pub struct RollingFileWriter<'a> {
+    inner: &'a Mutex<InnerAppender<BufWriterFactory>>,
+}
+
+impl<'a> RollingFileWriter<'a> {
+    fn new(inner: &'a Mutex<InnerAppender<BufWriterFactory>>) -> Self {
+        Self { inner }
+    }
+
+    fn inner(&self) -> io::Result<MutexGuard<'a, InnerAppender<BufWriterFactory>>> {
+        self.inner.lock().map_err(
+            |_: std::sync::PoisonError<MutexGuard<InnerAppender<BufWriterFactory>>>| {
+                io::Error::from(io::ErrorKind::Other)
+            },
+        )
+    }
+}
+
+impl<'a> io::Write for RollingFileWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.inner()?.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner()?.flush()
+    }
+}
+
+impl<'a> MakeWriter for RollingFileWriter<'a> {
+    type Writer = RollingFileWriter<'a>;
+
+    fn make_writer(&self) -> Self::Writer {
+        RollingFileWriter::new(&self.inner)
+    }
+}
+
+pub fn hourly<P: AsRef<Path>>(directory: P, file_name_prefix: P) -> RollingFileAppender {
```

Comment:
Being generic over a `P: AsRef<Path>` type and making both the directory _and_ prefix both be `P`s means that this can only be called with two instances of the _same_ type implementing `AsRef<Path>`. This means I can't, for example, write code like this: 
```rust
fn log_file_directory(/* ...  */) -> PathBuf { 
    /* ...  */
}
let log_dir = log_file_path(/* ... */);
let appender = tracing_appender::rolling::hourly(&log_dir, "my-app-logs-");
```
since `&PathBuf` and `&str` are different types, even though they both implement `AsRef<Path>`.

instead, if we wrote
```suggestion
pub fn hourly(directory: impl AsRef<Path>, file_name_prefix: impl AsRef<Path>) -> RollingFileAppender {
```
this could be called with any two types that implement `AsRef<Path>`. (note that we could also have two generic parameters, but that's more verbose.)

and, let's make the same change for `daily` and `never`.